packages/gossip/neighbor.go

```diff
@@ -33,149 +35,130 @@ const (
 	log = log.With(
 		"id", p.ID(),
-		"network", conn.LocalAddr().Network(),
-		"addr", conn.RemoteAddr().String(),
+		"localAddr", stream.Conn().LocalMultiaddr(),
+		"remoteAddr", stream.Conn().RemoteMultiaddr(),
 	)
-
 	return &Neighbor{
-		Peer:                  p,
-		Group:                 group,
-		BufferedConnection:    buffconn.NewBufferedConnection(conn, maxPacketSize),
-		log:                   log,
-		queue:                 make(chan []byte, neighborQueueSize),
-		closing:               make(chan struct{}),
-		connectionEstablished: time.Now(),
+		Peer:  p,
+		Group: group,
+
+		log: log,
+
+		stream: stream,
+		reader: libp2putil.NewDelimitedReader(stream),
+		writer: libp2putil.NewDelimitedWriter(stream),
+
+		disconnected:   events.NewEvent(disconnected),
+		packetReceived: events.NewEvent(packetReceived),
+		packetsRead:    atomic.NewUint64(0),
+		packetsWritten: atomic.NewUint64(0),
 	}
 }
 
-// ConnectionEstablished returns the connection established.
-func (n *Neighbor) ConnectionEstablished() time.Time {
-	return n.connectionEstablished
+// PacketsRead returns number of packets this neighbor has received.
+func (n *Neighbor) PacketsRead() uint64 {
+	return n.packetsRead.Load()
 }
 
-// Listen starts the communication to the neighbor.
-func (n *Neighbor) Listen() {
-	n.wg.Add(2)
-	go n.readLoop()
-	go n.writeLoop()
-
-	n.log.Info("Connection established")
+// PacketsWritten returns number of packets this neighbor has sent.
+func (n *Neighbor) PacketsWritten() uint64 {
+	return n.packetsWritten.Load()
 }
 
-// Close closes the connection to the neighbor and stops all communication.
-func (n *Neighbor) Close() error {
-	err := n.disconnect()
-	// wait for everything to finish
-	n.wg.Wait()
-
-	return err
+func disconnected(handler interface{}, _ ...interface{}) {
+	handler.(func())()
 }
 
-// IsOutbound returns true if the neighbor is an outbound neighbor.
-func (n *Neighbor) IsOutbound() bool {
-	return GetAddress(n.Peer) == n.RemoteAddr().String()
+func packetReceived(handler interface{}, params ...interface{}) {
+	handler.(func(*pb.Packet))(params[0].(*pb.Packet))
 }
 
-func (n *Neighbor) disconnect() (err error) {
-	n.disconnectOnce.Do(func() {
-		close(n.closing)
-		err = n.BufferedConnection.Close()
-		n.log.Info("Connection closed")
-	})
-	return
+// ConnectionEstablished returns the connection established.
+func (n *Neighbor) ConnectionEstablished() time.Time {
+	return n.stream.Stat().Opened
 }
 
-func (n *Neighbor) writeLoop() {
+func (n *Neighbor) readLoop() {
+	n.wg.Add(1)
 	defer n.wg.Done()
-
-	for {
-		select {
-		case msg := <-n.queue:
-			if len(msg) == 0 {
-				continue
-			}
-
-			n.log.Debugf("popped request from queue / remaining size %d: ", len(n.queue))
-
-			if _, err := n.BufferedConnection.Write(msg); err != nil {
-				n.log.Warnw("Write error", "err", err)
-				_ = n.disconnect()
+	go func() {
+		for {
+			packet := &pb.Packet{}
+			err := n.read(packet)
+			// the problem is here
+			if err != nil {
```

Comment:
Just for my understanding: this loop and therefore the waitgroup can only be terminated if `n.read()` returns an error. That means that when a different goroutine calls `n.disconnect()`, `n.read()` will execute this again which is only executed once because of the `sync.Once`?

```
if err != nil {
		disconnectErr := n.disconnect()
		return errors.CombineErrors(err, disconnectErr)
}
```

If that is the case and there's no better way of doing it, then I'd appreciate a clarifying comment here that explains how this whole thing terminates gracefully.

Reply:
Yes. Your understanding is correct. Sure, I will add a comment